rust/uefi/linux-bootloader/src/companions.rs

```diff
@@ -0,0 +1,158 @@
+    // and perform replacement of the last node before END_ENTIRE
+    // by another node containing the filename + .extra
+    // But this is as much tedious as performing a conversion to string
+    // then opening the root directory and finding the new directory.
+    let mut target_directory = loaded_image_file_path
+        .to_string(boot_services, DisplayOnly(false), AllowShortcuts(false))
+        // This is the Result-level error
+        .expect("Failed to obtain the string representation of the loaded image file path");
+    target_directory.push_str(cstr16!(".extra"));
+
+    return Ok(fs
+        .metadata(target_directory.as_ref())
+        .ok()
+        .and_then(|metadata| {
+            if metadata.is_directory() {
+                Some(PathBuf::from(target_directory))
+            } else {
+                None
+            }
+        }));
+}
+
+/// Potential companion initrd assembled on the fly
+/// during discovery workflows, e.g. finding files in drop-in directories.
+pub enum CompanionInitrd {
+    Credentials(Cpio),
+    GlobalCredentials(Cpio),
+    SystemExtension(Cpio),
+    PcrSignature(Cpio),
+    PcrPublicKey(Cpio),
+}
+
+impl CompanionInitrd {
+    pub fn into_cpio(self) -> Cpio {
+        match self {
+            Self::Credentials(c)
+            | Self::GlobalCredentials(c)
+            | Self::SystemExtension(c)
+            | Self::PcrSignature(c)
+            | Self::PcrPublicKey(c) => c,
+        }
+    }
+}
+
+/// Discover any credentials, i.e. files ending by .cred
+/// Credentials comes into two variants:
+///     - global credentials ($ESP/loader/credentials/*.cred), global to the ESP
+///     - local credentials ($path_to_image.extra/*.cred), specific to this image
+///
+/// Those will be unmeasured, you are responsible for measuring them or not.
+/// But CPIOs are guaranteed to be stable and independent of file discovery order.
+pub fn discover_credentials(
+    fs: &mut uefi::fs::FileSystem,
+    default_dropin_dir: Option<&Path>,
+) -> uefi::Result<Vec<CompanionInitrd>> {
+    let mut companions = Vec::new();
+
+    let default_global_dropin_dir = cstr16!("\\loader\\credentials");
+    if fs.try_exists(default_global_dropin_dir).unwrap() {
+        let metadata = fs.metadata(default_global_dropin_dir).expect("Failed to obtain metadata on `\\loader\\credentials` path (which is supposed to exist)");
+        if metadata.is_directory() {
+            let global_credentials: Vec<PathBuf> =
+                find_files(fs, default_global_dropin_dir.as_ref(), ".cred")?;
+
+            if !global_credentials.is_empty() {
+                companions.push(CompanionInitrd::GlobalCredentials(
+                    pack_cpio(
+                        fs,
+                        global_credentials,
+                        ".extra/global_credentials",
+                        0o500,
+                        0o400,
+                    )
+                    .expect("Failed to pack CPIO"),
```

Comment:
Should we convert the CPIO errors into uefi errors and return instead of panicking? Same below.

Reply:
I can return an LOAD_ERROR but not very sure if that's a good idea for debuggability given that our logging facilities are quite bad.